contracts/adapters/UniswapAdapter.sol

```diff
@@ -0,0 +1,233 @@
+    string[2] private venues = [_uniswapExchange, _sushiswapExchange];
+    address public immutable wbtcAddress;
+    bool private constant _nonFungible = true;
+    IUniswapV2Router02 public immutable uniswapRouter;
+    IUniswapV2Router02 public immutable sushiswapRouter;
+    IUniswapV2Pair public immutable wbtcDiggUniswap;
+    IUniswapV2Pair public immutable wbtcDiggSushiswap;
+    uint256 slippageDivisor = 500;
+
+
+   constructor(
+        address _uniswapRouter,
+        address _sushiswapRouter,
+        address _wbtcAddress,
+        address _ethAddress,
+        address _wbtcDiggUniswap,
+        address _wbtcDiggSushiswap,
+        address _diggAddress
+   ) public {
+
+        wbtcAddress = _wbtcAddress;
+        ethAddress = _ethAddress;
+        diggAddress = _diggAddress;
+        uniswapRouter = IUniswapV2Router02(_uniswapRouter);
+        sushiswapRouter = IUniswapV2Router02(_sushiswapRouter);
+        wbtcDiggUniswap = IUniswapV2Pair(_wbtcDiggUniswap);
+        wbtcDiggSushiswap = IUniswapV2Pair(_wbtcDiggSushiswap);
+
+    }
+
+   receive() external payable {}
+
+
+   function modifySlippage(uint256 newSlippageDivisor) onlyOwner public {
+        slippageDivisor = newSlippageDivisor;
+
+    }
+
+    function getSlippageDivisor() public view returns (uint256) {
+        return slippageDivisor;
+    }
+
+    function protocolName() public pure returns (string memory) {
+        return _name;
+    }
+
+    function nonFungible() external pure  returns (bool) {
+        return _nonFungible;
+    }
+
+    function sqrt(uint y) internal pure returns (uint z) {
+        if (y > 3) {
+            z = y;
+            uint x = y / 2 + 1;
+            while (x < z) {
+                z = x;
+                x = (y / x + x) / 2;
+            }
+        } else if (y != 0) {
+            z = 1;
+        }
+    }
+
+    // The core math involved in getting optimal swap amt to provide amm liquidity
+    function getSwapAmt(uint256 amtA, uint256 resA) internal pure returns (uint256){
+        return sqrt(amtA.mul(resA.mul(3988000) + amtA.mul(3988009))).sub(amtA.mul(1997))/1994;
+    }
+
+    function validateExchange(string memory exchangeName) internal view returns (bool){
+        for (uint i = 0; i < 2; i++){
+                if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(venues[i]))){
+                        return true;
+                }
+        }
+        return false;
+    }
+
+    function convertEthToToken( uint256 inputAmount, address addr, string memory exchangeName) internal {
+        if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_uniswapExchange))){
+                _convertEthToToken(inputAmount, addr, uniswapRouter);
+        }
+        else if (keccak256(abi.encodePacked(exchangeName)) == keccak256(abi.encodePacked(_sushiswapExchange))){
+                _convertEthToToken(inputAmount, addr, sushiswapRouter);
+        }

```

Comment:
We really should avoid using `keccak` as much as possible. One way we can avoid doing hash comparisons is just to use an enum to represent the exchanges as well:

```solidity
enum Exchange {Uniswap, Sushiswap}

function convertEthToToken(uint, address, Exchange exchange) {
  address router = exchange == Exchange.Uniswap ? uniswapRouter : sushiswapRouter);
}
```